#161290913 Add Search Functionality

ovieokeh opened the pull request:
#### What does this PR do?
This PR adds search functionality to the Authors Haven API. Search can be further refined using filters.
- Add search functionality
- Add ability to filter the search by author
- Add tests to ensure that the feature works as expected

#### Description of Task to be completed?
- `/POST 200 OK - /search`
    --- Required field: `keywords`

#### How should this be manually tested?
1. Clone the repo and run npm i
2. Create a `.env` file using the `.env.sample` example provided
3. Create a new user #11 and copy the `x-token` returned
4. Add as many articles as you want #13 using the copied `x-token` as authorization
5. Send a POST request to `/api/v1/search` and adding a search term in the `keywords`  field
6. If the search term entered matches an article in the DB, an array containing that article will be returned

##### Filtering by author
1. After adding your search term in the `keywords` field
2. Add `?author=[username]` to the URL where `[username]` is the username of the author you want to filter by (check the screenshots section for more information)
3. Now you will only get a result if the article with the search term was posted by the author being queried.

#### Any background context you want to provide?
The ability to filter search by article tags will be coming in the near future.

#### What are the relevant pivotal tracker stories?
#161290913

#### Screenshots (if appropriate)
<img width="1137" alt="screen shot 2018-11-03 at 11 57 46 pm" src="https://user-images.githubusercontent.com/36507972/47958099-50623e80-dfc4-11e8-9c3c-02bbaf416bb1.png">

<img width="1144" alt="screen shot 2018-11-04 at 12 03 27 am" src="https://user-images.githubusercontent.com/36507972/47958121-1e051100-dfc5-11e8-874b-cf40396f72af.png">

#### Questions:
N/A

oluseyianani committed:
feat(Authentication):user receive token on registration

     - issue user a valid token on successful registration
     - middleware to validate user token
     - TDD to check for token availability
     -user login functionality
     -user signup functionality

     [Delivers #161290897]

oluseyianani committed:
feat(Authentication):user receive token on registration

     - issue user a valid token on successful registration
     - middleware to validate user token
     - TDD to check for token availability
     -user login functionality
     -user signup functionality

     [Delivers #161290897]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(pagination): Pagination support for articles

        - Write test to check functionalities
        - Restructure folders and files tree
        - Modify routes folder, sequelizerc file to accommodate changes
        - Add validation for profile creation
        - Add validation for user login
        - Write test to test pagination features
        - Implement the feature proper
        - Test endpoint using postman
          [(Finishes) #161290910]

Commit:
feat(articles): create,  read, update and delete articles
- Write test for endpoints
- Implemented relationship in model and migration
- Validationhelpers for articles inputs
- Create article functionalities
- Update article functionalities
- Read article functionalities
- Delete article functionality

[Starts
 #161290903]

Commit:
feat(articles): create,  read, update and delete articles
- Write test for endpoints
- Implemented relationship in model and migration
- Validationhelpers for articles inputs
- Create article functionalities
- Update article functionalities
- Read article functionalities
- Delete article functionality

[Starts
 #161290903]

Commit:
feat(articles): create,  read, update and delete articles
    - Write test for endpoints
    - Implemented relationship in model and migration
    - Validationhelpers for articles inputs
    - Create article functionalities
    - Update article functionalities
    - Read article functionalities
    - Delete article functionality

    [Finishes #161290903]

Commit:
feat(articles): create,  read, update and delete articles
    - Write test for endpoints
    - Implemented relationship in model and migration
    - Validationhelpers for articles inputs
    - Create article functionalities
    - Update article functionalities
    - Read article functionalities
    - Delete article functionality

    [Finishes #161290903]

oluseyianani committed:
feat: list all user functionality

- unit test to check functionality
- list all user functionality
- validate user access with route protection

[Delivers #161290912]

oluseyianani committed:
feat: list all user functionality

- unit test to check functionality
- list all user functionality
- validate user access with route protection

[Delivers #161290912]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

oluseyianani commented:
Good job @65black. Can you try to make sure the coveralls pass? And I need to know if I can search for keywords like "and", "so", "or" e.t.c

ovieokeh commented:
@anasey01 Thanks for your review. All the code I added is all covered 100%. Maybe it's because I branched off a branch with fewer tests than the latest version.

ovieokeh committed:
Merge branch 'feature/161290910/pagination-support-for-articles' of https://github.com/andela/noldor-ah-backend into feature/161290910/pagination-support-for-articles

ovieokeh committed:
Merge branch 'feature/161290910/pagination-support-for-articles' of https://github.com/andela/noldor-ah-backend into feature/161290910/pagination-support-for-articles

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]

desaawa committed:
Merge pull request #12 from andela/feature/161290912/list-all-user-functionality

#161290912 List All Users Functionality

desaawa commented:
@65black Please resolve the conflict here.

ovieokeh committed:
Merge branch 'feature/161290913/search-functionality' of https://github.com/andela/noldor-ah-backend into feature/161290913/search-functionality

ovieokeh committed:
feat(search): add search functionality
- add index and triggers to articles table
- implement search by keywords alone
- implement search by both keywords and filters
- add tests for search feature
- add search inputs validation
[Finishes #161290903]
